homeassistant/components/recorder/__init__.py

```diff
@@ -823,10 +833,10 @@ def _async_setup_periodic_tasks(self):
             self.hass, self.async_periodic_statistics, minute=range(0, 60, 5), second=10
         )
 
-    def run(self):
+    def run(self) -> None:
         """Start processing events to save."""
         shutdown_task = object()
-        hass_started = concurrent.futures.Future()
+        hass_started: concurrent.futures.Future = concurrent.futures.Future()
```

Comment:
Is this future threadsafe? It seems we're abusing the `concurrent.futures.Future` api. Some of it is only meant for use in executor implementations:
https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future

Reply:
That uses a `threading.Condition` under the hood for access so it should be thread-safe.  While it is not broken AFAICT, it would probably be good to refactor the implementation at some point as I think there are better options available now.

Reply:
Ok.

Reply:
Looks like asyncio.run_coroutine_threadsafe does as well

Reply:
Either one is still probably a better option here since at least it means if something does change with it there will be less places to fix

Reply:
It's late late here so something to think about for tomorrow.

Reply:
Need to do more testing, but https://github.com/home-assistant/core/pull/68593